test/distributed/fsdp/test_fsdp_comm_hooks.py

```diff
@@ -0,0 +1,239 @@
+# Owner(s): ["oncall: distributed"]
+
+import copy
+import sys
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import distributed as dist
+from torch.distributed.algorithms._comm_hooks import allreduce_hook
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import ShardingStrategy
+
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+class TestCommunicationHooks(FSDPTest):
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "sharding_strategy",
+        [
+            ShardingStrategy.NO_SHARD
+        ])
+    def test_default_communication_hook_behaviour(
+        self,
+        sharding_strategy: Optional[ShardingStrategy]
+    ):
+        """
+        Tests FSDP's default communication hook's behaviour and correctness.
+        Arguments:
+            has_wrapping (bool): Configures wrapping of a module.
+            sharding_strategy (Optional[ShardingStrategy]): Configures the FSDP algorithm.
+        """
+        m = torch.nn.Linear(1, 2, bias=False)
+        inpt = torch.tensor([self.rank]).float().cuda(self.rank)
+
+        net_default_hook = FSDP(
+            copy.deepcopy(m),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy
+        ).to(self.rank)
+
+        # Check that default hook is set to `all_reduce`
+        for entry in FSDP.fsdp_modules(net_default_hook):
+            self.assertEqual(entry.communication_hook.__qualname__, allreduce_hook.allreduce_hook.__qualname__)
+
+        for _ in range(4):
+
+            # Clear gradients manually.
+            for entry in net_default_hook.parameters():
+                if entry.grad is not None:
+                    entry.grad.requires_grad_(False)
+                    entry.grad.zero_()
+
+            loss = net_default_hook(inpt).sum()
+            loss.backward()
+
+            # For each worker, the gradient on the weight should be worker_rank.
+            grad = [param.grad for param in net_default_hook.parameters()]
+
+            avg = copy.deepcopy(grad[0])
+            expected_grad = (
+                sum(i for i in range(dist.get_world_size())) / dist.get_world_size()
+            )
+            # Verify default hook produces expected gradients
+            self.assertEqual(
+                avg[0].item(),
```

Comment:
noob question: Why are we taking `avg[0].item()` here? I may be misreading, but it looks like `grad = [flat_param.grad]` and `avg = grad[0] = flat_param.grad`. Does this mean that `avg[0]` is the 0th element of `avg`, returned as a `Tensor`? Is there any need to call `avg[0].item()`, or can we just call `avg.item()`?

Reply:
`avg.item()` won't work, because `avg` has multiple elements. Test works in such a way, that gradient values are the same, i.e. in this case for 1X2 network `grad` would be `tensor([0.5000, 0.5000])`. Thus, we only need to compare one of these values with the expected one.